mono/profiler/arg-parsing.c

```diff
@@ -0,0 +1,406 @@
+	if (!val)
+		return;
+	if (strcmp (val, "ondemand") == 0) {
+		config->hs_mode_ondemand = TRUE;
+		return;
+	}
+
+	count = strtoul (val, &end, 10);
+	if (val == end) {
+		usage ();
+		return;
+	}
+
+	if (strcmp (end, "ms") == 0)
+		config->hs_mode_ms = count;
+	else if (strcmp (end, "gc") == 0)
+		config->hs_mode_gc = count;
+	else
+		usage ();
+}
+
+static void
+set_sample_freq (ProfilerConfig *config, const char *val)
+{
+	if (!val)
+		return;
+
+	const char *p = val;
+
+	// Is it only the frequency (new option style)?
+	if (isdigit (*p))
+		goto parse;
+
+	// Skip the sample type for backwards compatibility.
+	while (isalpha (*p))
+		p++;
+
+	// Skip the forward slash only if we got a sample type.
+	if (p != val && *p == '/') {
+		p++;
+
+		char *end;
+
+	parse:
+		config->sample_freq = strtoul (p, &end, 10);
+
+		if (p == end) {
+			usage ();
+			return;	
+		}
+
+		p = end;
+	}
+
+	if (*p)
+		usage ();
+}
+
+static void
+usage (void)
+{
+	printf ("Log profiler version %d.%d (format: %d)\n", LOG_VERSION_MAJOR, LOG_VERSION_MINOR, LOG_DATA_VERSION);
+	printf ("Usage: mono --profile=log[:OPTION1[,OPTION2...]] program.exe\n");
+	printf ("Options:\n");
+	printf ("\thelp                 show this usage info\n");
+	printf ("\t[no]'event'          enable/disable a profiling event. Valid values: domain, assembly, module, class, jit, exception, gcalloc, gc, thread, monitor, gcmove, gcroot, context, finalization, counter, gchandle\n");
+	printf ("\t[no]typesystem       enable/disable typesystem related events such as class and assembly loading\n");
+	printf ("\t[no]alloc            enable/disable recording allocation info\n");
+	printf ("\t[no]calls            enable/disable recording enter/leave method events\n");
+	printf ("\t[no]legacy           enable/disable pre mono 5.4 default profiler events\n");
+	printf ("\tsample[=TYPE]        enable/disable statistical sampling of threads (by default cycles/100)\n");
+	printf ("\t                     TYPE: cycles,instr,cacherefs,cachemiss,branches,branchmiss\n");
+	printf ("\t                     TYPE can be followed by /FREQUENCY\n");
+	printf ("\t[heapshot[=MODE]     record heap shot info (by default at each major collection)\n");
+	printf ("\t                     MODE: every XXms milliseconds, every YYgc collections, ondemand\n");
+	printf ("\t[no]coverage         enable collection of code coverage data\n");
+	printf ("\tcovfilter=ASSEMBLY   add an assembly to the code coverage filters\n");
+	printf ("\t                     add a + to include the assembly or a - to exclude it\n");
+	printf ("\t                     filter=-mscorlib\n");
+	printf ("\tcovfilter-file=FILE  use FILE to generate the list of assemblies to be filtered\n");
+	printf ("\tmaxframes=NUM        collect up to NUM stack frames\n");
+	printf ("\tcalldepth=NUM        ignore method events for call chain depth bigger than NUM\n");
+	printf ("\toutput=FILENAME      write the data to file FILENAME (-FILENAME to overwrite)\n");
```

Comment:
We always overwrite now, so we should probably take this opportunity to drop support for `-FILENAME`, or at least deprecate it.